fs/rails-base-graphql-api
# Use Shrine for image uploading

### Summary

- added gems 'aws-sdk-s3' and 'shrine'
- removed ActiveStorage
- added GraphQL mutation to get URL and params for direct upload to AWS S3

This PR implements the first point of the next image uploading flow:
- Client fetches the presign information from the server, and use this information to upload the file to S3
- Client uploads file directly to S3
- After uploading client generates a JSON representation of the uploaded file and sends it with other data to the server
- Once data on the server updated the background job is kicked off to process image
- The record is saved with cached file, which is shown as fallback
- Once background job is finished, record is updated with processed attachment data

### How it works

![Screen Shot 2020-06-19 at 17 37 51](https://user-images.githubusercontent.com/58729845/85144374-beb84b00-b253-11ea-8302-aa96d28bd1b1.png)

### Test plan

List of steps to manually test introduced functionality:

* Fill `.env` file with AWS S3 credentials
* Open GraphiQL App
* Make request using schema:
```
mutation Presign($filename: String!, $type: String!) {
  presignData(filename: $filename, type: $type) {
    url
    headers
    fields
  }
}
```
* Check that response has URL for upload and fields that FE can use as upload params

### Review notes

While reviewing pull-request (especially when it's your pull-request),
please make sure that:

- you understand what problem is solved by PR and how is it solved
- new tests are in place, no redundant tests
- DB schema changes reflect new migrations
- newly introduced DB fields have indexes and constraints
- there are no missed files (migrations, view templates)
- required ENV variables added and described in `.env.example` and added to Heroku
- associated Heroku review app works correctly with introduced changes

### Deploy notes

Before deploy the next environment variables should be added:
- S3_BUCKET_NAME
- S3_BUCKET_REGION
- S3_ACCESS_KEY_ID
- S3_SECRET_ACCESS_KEY

### References
* Github Issue #42
